core/trino-main/src/main/java/io/trino/server/GracefulShutdownWorkerHandler.java

```diff
@@ -0,0 +1,153 @@
+    private final LifeCycleManager lifeCycleManager;
+    private final SqlTaskManager sqlTaskManager;
+    private final ShutdownAction shutdownAction;
+    private final Duration gracePeriod;
+
+    @GuardedBy("this")
+    private boolean shutdownRequested;
+
+    @Inject
+    public GracefulShutdownWorkerHandler(
+            SqlTaskManager sqlTaskManager,
+            ServerConfig serverConfig,
+            ShutdownAction shutdownAction,
+            LifeCycleManager lifeCycleManager)
+    {
+        this.sqlTaskManager = requireNonNull(sqlTaskManager, "sqlTaskManager is null");
+        this.shutdownAction = requireNonNull(shutdownAction, "shutdownAction is null");
+        this.lifeCycleManager = requireNonNull(lifeCycleManager, "lifeCycleManager is null");
+        this.gracePeriod = requireNonNull(serverConfig, "serverConfig is null").getGracePeriod();
+    }
+
+    @Override
+    public synchronized void requestShutdown()
+    {
+        log.info("Shutdown requested");
+        if (shutdownRequested) {
+            return;
+        }
+        shutdownRequested = true;
+
+        // wait for a grace period (so that shutting down state is observed by the coordinator) to start the shutdown sequence
+        shutdownHandler.schedule(this::shutdown, gracePeriod.toMillis(), MILLISECONDS);
+    }
+
+    private void shutdown()
+    {
+        List<TaskInfo> activeTasks = getActiveTasks();
+
+        // At this point no new tasks should be scheduled by coordinator on this worker node.
+        // Wait for all remaining tasks to finish.
+        while (activeTasks.size() > 0) {
+            CountDownLatch countDownLatch = new CountDownLatch(activeTasks.size());
+
+            for (TaskInfo taskInfo : activeTasks) {
+                sqlTaskManager.addStateChangeListener(taskInfo.getTaskStatus().getTaskId(), newState -> {
+                    if (newState.isDone()) {
+                        countDownLatch.countDown();
+                    }
+                });
+            }
+
+            log.info("Waiting for all tasks to finish");
+
+            try {
+                countDownLatch.await();
+            }
+            catch (InterruptedException e) {
+                log.warn("Interrupted while waiting for all tasks to finish");
+                currentThread().interrupt();
+            }
+
+            activeTasks = getActiveTasks();
+        }
+
+        // wait for another grace period for all task states to be observed by the coordinator
+        sleepUninterruptibly(gracePeriod.toMillis(), MILLISECONDS);
+
+        Future<?> shutdownFuture = lifeCycleStopper.submit(() -> {
+            lifeCycleManager.stop();
+            return null;
+        });
+
+        // terminate the jvm if life cycle cannot be stopped in a timely manner
+        try {
+            shutdownFuture.get(LIFECYCLE_STOP_TIMEOUT.toMillis(), MILLISECONDS);
```

Comment:
I thought the outside task is doing the same. It calls graceful shutdown and waits till completion for some time, if it is still up and running it kills-nine it